src/server/reply.rs

```diff
@@ -1,37 +1,95 @@
 use serde::Serialize;
+use warp::http::header::HeaderValue;
 use warp::http::status::StatusCode;
 use warp::reply::Response;
 use warp::Reply;
 
-use super::TOML_MIME_TYPE;
+use tracing::log::debug;
+
+use super::{JSON_MIME_TYPE, TOML_MIME_TYPE};
 use crate::provider::ProviderError;
 
-// Borrowed and modified from https://docs.rs/warp/0.2.5/src/warp/reply.rs.html#102
-pub fn toml<T>(val: &T) -> Toml
+/// Use an accept header to determine how to serialize content.
+///
+/// This will examine the Accept header, looking for the best match, and then it will
+/// use the appropriate serializer to serialize the data.
+///
+/// The current implementation ignores `q=` annotations, assigning preference based on
+/// the first MIME type to match.
+///
+/// For example, `Accept: text/json, application/toml;q=0.9` will cause encoding to be in JSON.
+/// If no suitable content type is found, this will encode in application/toml, as that
+/// is the behavior described in the spec.
+pub fn serialized_data<T>(val: &T, accept: String) -> SerializedData
 where
     T: Serialize,
 {
-    Toml {
-        inner: toml::to_vec(val).map_err(|e| {
+    let supported = vec![
+        TOML_MIME_TYPE.to_owned(),
+        JSON_MIME_TYPE.to_owned(),
+        "text/json".to_owned(),
+    ];
+    let default_mime = "*/*".to_owned();
+    let accept_items = parse_accept(accept.clone());
+    debug!(
+        "Parsed accept header '{}' into items: {:?}",
+        accept, accept_items
+    );
+    let best_fit = accept_items
+        .iter()
+        .find(|i| supported.contains(*i))
+        .unwrap_or(&default_mime);
+    debug!("Best fit MIME is {}", best_fit);
+    let mut final_mime = TOML_MIME_TYPE;
+    let inner = match (*best_fit).as_str() {
+        JSON_MIME_TYPE | "text/json" => {
+            final_mime = JSON_MIME_TYPE;
+            serde_json::to_vec(val).map_err(|e| {
+                tracing::log::error!("Error while serializing TOML: {:?}", e);
+            })
+        }
+        // TOML is default
+        _ => toml::to_vec(val).map_err(|e| {
             tracing::log::error!("Error while serializing TOML: {:?}", e);
         }),
+    };
+    debug!("negotiated MIME is '{}'", final_mime);
+    SerializedData {
+        inner,
+        mime: final_mime.to_owned(),
     }
 }
 
-/// A JSON formatted reply.
-pub struct Toml {
+fn parse_accept<'a>(header: String) -> Vec<String> {
```

Comment:
Nit: This doesn't need to be a `String` as `split` is a method from `&str`. It will avoid an extra allocation as well as you already clone the parts you need